File: Snail.cpp
#include "Snail.hpp"
#include "Aquarium.hpp"

Snail::Snail(Aquarium& _aquarium) : Creature(_aquarium, 3, Vector2((double)_aquarium.getSizeX() / 2, _aquarium.getSizeY()), 50) {
    
}

bool Snail::operator==(const Snail& other) const {
    return this == &other;
}

void Snail::move() {
    Vector2 coinPosition = eat();
    if (coinPosition != Vector2::null) {
        double fX = coinPosition.x;
        double sX = getPosition().x;
        Vector2 direction = Vector2(fabs(fX - sX) < 0.05 * getSpeed() ? 0 : fX > sX ? 1 : -1, 0);
        setPosition(getPosition() + direction * getSpeed());
        setIsMovingRight(direction.x >= 0);
    } 
}

Vector2 Snail::eat() {
    Coin* coin = findCoin();
    if (coin) {
        Vector2 coinPosition = coin->getPosition();
        for (ElementList<Coin>* e = getAquarium().getCoins().getFirst(); e != 0; e = e->next) {
            if (getPosition().distance(e->data.getPosition()) <= getEatRadius()) {
                e->data.take();
            }
        }

        return coinPosition;
    } else {
        return Vector2::null;
    }
}

Coin* Snail::findCoin() const {
    LinkedList<Coin>& coins = getAquarium().getCoins();
    if (coins.isEmpty()) {
        return 0;
    } else {
        struct Priority {
            bool onFloor;
            double distance;
        };
        double aquariumSizeY = getAquarium().getSizeY();
        ElementList<Coin>* coin = coins.getFirst();
        Vector2 coinPosition = coin->data.getPosition();
        
        Priority coinPriority;
        if (coinPosition.y == aquariumSizeY) {
            coinPriority.onFloor = true;
            coinPriority.distance = fabs(coinPosition.x - getPosition().x);
        } else {
            coinPriority.onFloor = false;
            coinPriority.distance = aquariumSizeY - coinPosition.y;
        }

        for (ElementList<Coin>* e = coin->next; e != 0; e = e->next) {
            Vector2 ePosition = e->data.getPosition();

            if (coinPriority.onFloor) {
                if (ePosition.y == aquariumSizeY) {
                    double eDistance = fabs(ePosition.x - getPosition().x);
                    if (eDistance < coinPriority.distance) {
                        coin = e;
                        coinPriority.distance = eDistance;
                    }
                }
            } else {
                if (ePosition.y == aquariumSizeY) {
                    coin = e;
                    coinPriority.onFloor = true;
                    coinPriority.distance = fabs(ePosition.x - getPosition().x);
                } else {
                    double eDistance = aquariumSizeY - ePosition.y;
                    if (eDistance < coinPriority.distance) {
                        coin = e;
                        coinPriority.distance = eDistance;
                    }
                }
            }
        }

        return &coin->data;
    }
}

File: production/Java/src/Snail.java
public class Snail extends Creature {

  /**
	 * Constructor: instantiate snail on the bottom middle of the aquarium, speed of 3, and eatRadius of 50
	 * @param aquarium object aquarium which will be added by piranha and will be initialized
	 */
  public Snail(Aquarium aquarium) {
    super(aquarium, 3, new Vector2((double) aquarium.getSizeX() / 2, aquarium.getSizeY()), 50);
  }

  /**
     * Reference comparison
     * @param other other snail to compare
     * @return true if and only if argument represent this snail
     */
  public boolean equals(final Snail other) {
    return this == other;
  }

  /**
     * Move horizontally
     */
  public void move() {
    Vector2 coinPosition = eat();
    if (coinPosition != null) {
      double foodX = coinPosition.abscissa;
      double snailX = getPosition().abscissa;
      Vector2 direction = new Vector2(Double.compare(foodX, snailX), 0);
      setPosition(getPosition().add(direction.multiply(getSpeed())));
      setIsMovingRight(direction.abscissa >= 0);
    }
  }

  /**
     * Find coin in the aquarium
     * @return coin's position
     */
  public Vector2 eat() {
    Coin coin = findCoin();
    if (coin != null) {
      Vector2 coinPosition = coin.getPosition();
      for (ElementList<Coin> e = getAquarium().getCoins().getFirst(); e != null; e = e.getNext()) {
        if (getPosition().distance(e.getData().getPosition()) <= getEatRadius()) {
          e.getData().take();
        }
      }
      return coinPosition;
    } else {
      return null;
    }
  }

  /**
     * Get coin which will be get by snail
     * @return coin's data
     */
  private Coin findCoin() {
    LinkedList<Coin> coins = getAquarium().getCoins();
    if (coins.isEmpty()) {
      return null;
    } else {
      class Priority {
        boolean onFloor;
        double distance;
      }

      double aquariumSizeY = getAquarium().getSizeY();
      ElementList<Coin> coin = coins.getFirst();
      Vector2 coinPosition = coin.getData().getPosition();

      Priority coinPriority = new Priority();
      if (coinPosition.ordinate == aquariumSizeY) {
        coinPriority.onFloor = true;
        coinPriority.distance = Math.abs(coinPosition.abscissa - getPosition().abscissa);
      } else {
        coinPriority.onFloor = false;
        coinPriority.distance = aquariumSizeY - coinPosition.ordinate;
      }

      for (ElementList<Coin> e = coin.getNext(); e != null; e = e.getNext()) {
        Vector2 elementPosition = e.getData().getPosition();

        if (coinPriority.onFloor) {
          if (elementPosition.ordinate == aquariumSizeY) {
            double elementDistance = Math.abs(elementPosition.abscissa - getPosition().abscissa);
            if (elementDistance < coinPriority.distance) {
              coin = e;
              coinPriority.distance = elementDistance;
            }
          }
        } else {
          if (elementPosition.ordinate == aquariumSizeY) {
            coin = e;
            coinPriority.onFloor = true;
            coinPriority.distance = Math.abs(elementPosition.abscissa - getPosition().abscissa);
          } else {
            double elementDistance = aquariumSizeY - elementPosition.ordinate;
            if (elementDistance < coinPriority.distance) {
              coin = e;
              coinPriority.distance = elementDistance;
            }
          }
        }
      }
      return coin.getData();
    }
  }
}
